myapp/utility.py

```diff
@@ -1,116 +1,880 @@
+        professors = [p.username for p in c.users if (p.role_name == 'Admin' or p.role_name == 'SuperAdmin')]
+        students = [s.username for s in c.users if s.role_name == 'Student']
+        labs = [lab.id for lab in c.labs]
+        current_classes.append({'id': c.id,
+                                'name': c.name,
+                                'time': c.time,
+                                'professors': professors,
+                                'students': students,
+                                'labs': labs})
+    return current_classes
+
+
+# check the data sent by client and
+# deserialize the information of a lab from jsonData sent by client
+# and serialize it into JSON format dictionary that is consistent with
+# python side naming convention
+def deserialize_lab(lab_json):
+    err_msg = check_existence(lab_json, 'labName', 'className', 'classTime',
+                                        'professorName', 'labDescription',
+                                        'experiments')
+    if err_msg != '':
+        return (dict(), err_msg)
+
+    lab_info = {'name': lab_json['labName'],
+                'class_name': lab_json['className'],
+                'class_time': lab_json['classTime'],
+                'prof_name': lab_json['professorName'],
+                'description': lab_json['labDescription'],
+                'experiments': lab_json['experiments'],
+                'old_lab_id': lab_json['oldLabId']}
+
+    for e in lab_json['experiments']:
+        err_msg = check_existence(e, 'name', 'description', 'order',
+                                     'valueType', 'valueRange',
+                                     'valueCandidates')
+        if err_msg != '':
+            return (lab_info, err_msg)
+    return (lab_info, '')
+
+
+# Covert a string in the format '[a,b,c]' into a python list
+def deserialize_lab_id(lab_ids):
+    lab_ids = lab_ids.lstrip('\"[').rstrip('\"]').split(',')
+    lab_ids = [lab.lstrip('\'').rstrip('\'') for lab in lab_ids if
+               lab.lstrip('\'').rstrip('\'') != '']
+    if not(isinstance(lab_ids, list)):
+        err_msg = 'The <lab_ids> does not have the right format. '
+        +'lab_ids has the type '+str(type(lab_ids))
+        return lab_ids, err_msg
+    return lab_ids, ''
 
-# Check the existence of args in a form
-def Check_existence(form, *args):
-	err_msg = ''
-	for a in args:
-		if not(a in form):
-			err_msg += 'You must send a JSON object '+a+'\n'
-	return err_msg  
 
+# --- Find the information of labs ---
+# return all the list of labs with basic information according to
+# the current user's role
+def find_lab_list_for_user(user):
+    lab_list = []
+    # query labs according to the user's role
+    labs_query = query_labs_for_user(user)
+    if user.role_name == 'Admin':
+        labs_query = user.labs
+
+    # get all the info of the labs
+    for lab in labs_query:
+        data_num = 0
+        experiments_query = query_experiments_for_lab(lab.id)
```

Comment:
I'm a little worried that your convenience methods are both making this less clear and de-optimizing the django ORM, which is usually successful in keeping your database interactions pretty efficient.